### PR TITLE
Added support for setting certain oidc providers.

### PR DESCRIPTION
**Pull request by Sondre-B**
## Description
Adding support for Maskinportentokens by adding settings-classes for retrieving oidc provider settings and backwards-compability using PlatformSettings.cs.

## Related Issue(s)
- #187 - Add support for maskinporten token in JWTOCookie package (from altinn-accessmanagement repo)

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs [hunk 1]
```diff
@@ -179,5 +216,23 @@ protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
                 throw;
             }
         }
+
+        /// <summary>
+        /// Get the signing keys published by the given endpoint.
+        /// </summary>
+        /// <param name="url">The url of the endpoint</param>
+        /// <returns>The signing keys published by the endpoint</returns>
+        public async Task<ICollection<SecurityKey>> GetSigningKeys(string url)
```
Comment (TheTechArch) [inline src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs:225]: @SandGrainOne du drev vel litt å se på dette. Hvordan er det med caching her. Vil det være en innebygget caching her som gjør at man ikke henter signings key for samme hver gang?

File: src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs [hunk 2]
```diff
@@ -19,20 +22,30 @@ namespace AltinnCore.Authentication.JwtCookie
...
         public JwtCookieHandler(
             IOptionsMonitor<JwtCookieOptions> options,
+            IOptions<OidcProviderSettings> oidcProviderSettings,
+            IOptions<PlatformSettings> platformSettings,
             ILoggerFactory logger,
             UrlEncoder encoder,
             ISystemClock clock)
             : base(options, logger, encoder, clock)
         {
+            _oidcProviderSettings = oidcProviderSettings.Value;
```
Comment (jonkjetiloye) [inline src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs:47]: Why do we need both platformSettings and oidcProviderSettings?
PlatformSettings is the current config setup for AccessManagement right?

Shouldn't access management (and anyone else who is to use new version of the JWTCookie package simply have to update their config for the new OidcProviderSettings?

File: src/jwtcookie/Authentication/Altinn.Common.Authentication.csproj [hunk 3]
```diff
@@ -31,7 +31,7 @@
     <FrameworkReference Include="Microsoft.AspNetCore.App" />
     <PackageReference Include="Microsoft.Azure.KeyVault" Version="3.0.5" />
     <PackageReference Include="Microsoft.IdentityModel.Clients.ActiveDirectory" Version="5.3.0" />
-    <PackageReference Include="Microsoft.IdentityModel.Protocols.OpenIdConnect" Version="6.25.1" />
+    <PackageReference Include="Microsoft.IdentityModel.Protocols.OpenIdConnect" Version="6.25.0" />
```
Comment (jonkjetiloye) [inline src/jwtcookie/Authentication/Altinn.Common.Authentication.csproj:34]: Downgrade of patch version for OpenIdConnect? Might as well get latest so we don't have to patch this again on thursday

File: src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs (hunk 1 above)
Comment (SandGrainOne) [reply src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs:225]: Jeg mener det er innebygget caching her et sted. Fant dette dere kan nøyste videre i:
https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.tokens.baseconfigurationmanager.refreshinterval?view=azure-dotnet#microsoft-identitymodel-tokens-baseconfigurationmanager-refreshinterval

File: src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs [hunk 4]
```diff
@@ -19,20 +22,30 @@ namespace AltinnCore.Authentication.JwtCookie
...
+            _platformSettings = platformSettings.Value;
```
Comment (TheTechArch) [inline src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs:48]: I know I have thought about platform settings before I remembered it was configured directly. Do we really need it now?

File: src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs (hunk 2 above)
Comment (SandGrainOne) [reply src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs:47]: Not sure how this plays into it, but we also need to remember JwtCookieOptions. First of all the MetadataAddress property.

Comment (Sondre-B) [reply src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs:44]: I have removed the PlatformSettings class in the latest commit.

File: src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs [hunk 5]
```diff
@@ -101,9 +111,22 @@ protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
...
+                                    validationParameters.IssuerSigningKeys = await GetSigningKeys(provider.Value.WellKnownConfigEndpoint);
+                                }
+                            }
+                        }
+                        else
+                        {
+                            validationParameters.IssuerSigningKeys = validationParameters.IssuerSigningKeys?.Concat(configuration.SigningKeys) ?? configuration.SigningKeys;
+                        }
                     }
                 }
```
Comment (TheTechArch) [inline src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs:132]: Should we fallback to the default if there is no match in the loop for issuer

File: src/jwtcookie/Authentication/Altinn.Common.Authentication.csproj [hunk 6]
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <TargetFramework>net6.0</TargetFramework>
     <OutputType>Library</OutputType>
     <AssemblyVersion>2.4.2.0</AssemblyVersion>
```
Comment (TheTechArch) [inline src/jwtcookie/Authentication/Altinn.Common.Authentication.csproj:6]: Update nuget one major version since it is a .net core change.

Suggest 3.0.0

Comment (TheTechArch) [reply src/jwtcookie/Authentication/Altinn.Common.Authentication.csproj:6]: You can publish beta version of nuget if you need to test

File: src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs (hunk 5 above)
Comment (Sondre-B) [reply src/jwtcookie/Authentication/JwtCookie/JwtCookieHandler.cs:137]: @TheTechArch  Updated this piece of the code, is this what you mean?

**Review by jonkjetiloye — APPROVED** [review]

**Comment by TheTechArch** [comment]
Will merge when 3.0.0 non alpha is created

**Pull request merged**